python3.pkgs.ninja-python: replace with a stub implementation

tjni opened the pull request:
## Description of changes

This maps to the ninja package on PyPI, which downloads ninja from the web. This was packaged in https://github.com/NixOS/nixpkgs/pull/243967.

@999eagle, @SuperSandro2000 I would like to replace that implementation with a stub that directly uses ninja in nixpkgs. This capability was requested by @FRidh in https://github.com/scikit-build/ninja-python-distributions/issues/127 – however upstream would like to delegate this capability to PEP 517 backends.

Only certain backends support this out-of-the-box, such as scikit-build-core and (maybe) meson-python. Notably, some projects are still using the older scikit-build, which does not support this. This impacts https://github.com/NixOS/nixpkgs/pull/245509 because the way I have set up [`build`](https://pypa-build.readthedocs.io/en/latest/) does dependency validation and those projects either need to include this ninja package (picking up its version of ninja) or we need to patch their pyproject.toml.

Hence, I'd like to stub out this package and converge on using just the single ninja program provided in nixpkgs.

If this sounds reasonable, I'd also like some opinions on whether it's worth changing the name of this package to just ninja. On the one hand, duplicate names can be confusing. On the other hand, it could also be confusing to someone building a Python package and reading the pyproject.toml that they need to use ninja-python instead of simply ninja.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

tjni committed:
python3.pkgs.ninja-python: replace with a stub implementation

Instead of packaging the upstream package, which downloads Ninja from
the web, we can stub it out to use ninja from nixpkgs.

tjni committed:
python3.pkgs.ninja-python: add tjni as a maintainer of this stub

FRidh commented on pkgs/development/python-modules/ninja/stub/ninja/__init__.py:
```diff
@@ -0,0 +1,8 @@
+    return subprocess.call([name] + args, close_fds=False)
+
+def ninja():
+    raise SystemExit(_program('ninja', sys.argv[1:]))
```
Should we hardcode the path to the ninja executable so we get a proper runtime dependency

tjni replied:
I'm a little out of my depth with this. With normal derivations, I'm under impression that including in `buildInputs` is preferred since it's structured (maybe helps with cross too?) and will be added to PATH. With Python derivations, I don't know how it works exactly, but I had to add it to propagatedBuildInputs to achieve the same effect. What's the difference between doing that and hardcoding?

tjni replied:
I thought about it a bit more and like the idea of hardcoding the path in the function. I had to keep `ninja` also in propagatedBuildInputs because some users of this package do expect ninja to just be in PATH.

FRidh replied:
With regular derivations we do not need to explicitly hardcode because references get compiled in. With Python that does not happen so we need to explicitly hardcode.

tjni committed:
python3.pkgs.ninja: rename from ninja-python

I am hoping this can be a transparent shim for most if not all Python
packages that acts as a proxy to the actual ninja package in nixpkgs.
If this works, doing it this way simplifies rolling out the new Python
build hooks that do stricter build dependency validation.

tjni commented:
I also went ahead and optimistically renamed `ninja-python` to `ninja` so it can be a transparent shim to Python packages. My sense is that is how most Python packages receive ninja and, if this works, it will require less work to get dependency validation passing in the new Python bootstrapping process.

tjni commented:
I think I figured out how to do this properly. We don't need to propagate the original tool in the PATH. We only need to:

1. Define entry points that call into the stubs which wrap the tools. I didn't realize initially that this was how it worked.
2. Inherit the setup hooks from the parent package.

tjni committed:
python3.pkgs.meson-python: depend on top-level ninja

This is a build tool that depends on a ninja binary directly, not
through the ninja PyPI package.

bryango commented:
I believe python2 is broken by this PR; it lands on master via https://github.com/NixOS/nixpkgs/pull/248496, and I can get python2 to build on [its parent](https://github.com/NixOS/nixpkgs/commit/bceafd3a3ad0494c87922993a271606acabccf3d) but afterward, not anymore. Maybe it's time to give up python2 and mark it as broken...

tjni commented:
Can you share the command you're running? I will try to build Python 2 in a little bit, but the reason I ask is because this doesn't seem connected to Python 2 itself but perhaps some package that is using it? Were you building `nixops` and it passed before but fails after?

FRidh commented:
> I believe python2 is broken by this PR; it lands on master via #248496, and I can get python2 to build on [its parent](https://github.com/NixOS/nixpkgs/commit/bceafd3a3ad0494c87922993a271606acabccf3d) but afterward, not anymore. Maybe it's time to give up python2 and mark it as broken...

It builds fine on both master and staging-next.

bryango commented:
Hey sorry for the lack of details; the direct failure is always from `flit-core`; I was not building `nixops` myself but actually `gimp` with python2, and observe almost identical errors; [build log](https://github.com/bryango/nixpkgs-config/actions/runs/5971762400/job/16201291689) excerpt:
```
error: builder for '/nix/store/1px104l2nbjb0v88fzcqdk21w2l0k1cq-python2.7-flit-core-3.9.0.drv' failed with exit code 2;
       last 10 log lines:
       >     backend=self.pep517_backend,
       >   File "/nix/store/vm2917jk0jsbanpkl9z74fflz1achliq-python2.7-pip-20.3.4/lib/python2.7/site-packages/pip/_internal/operations/build/metadata.py", line 35, in generate_metadata
       >     metadata_dir
       >   File "/nix/store/vm2917jk0jsbanpkl9z74fflz1achliq-python2.7-pip-20.3.4/lib/python2.7/site-packages/pip/_vendor/pep517/wrappers.py", line 196, in prepare_metadata_for_build_wheel
       >     '_allow_fallback': _allow_fallback,
       >   File "/nix/store/vm2917jk0jsbanpkl9z74fflz1achliq-python2.7-pip-20.3.4/lib/python2.7/site-packages/pip/_vendor/pep517/wrappers.py", line 284, in _call_hook
       >     raise BackendUnavailable(data.get('traceback', ''))
       > BackendUnavailable
       > Removed file:///build/source/flit_core from build tracker '/build/pip-req-tracker-M2uOIa'
       > Removed build tracker: '/build/pip-req-tracker-M2uOIa'
       For full logs, run 'nix log /nix/store/1px104l2nbjb0v88fzcqdk21w2l0k1cq-python2.7-flit-core-3.9.0.drv'.
error: 1 dependencies of derivation '/nix/store/xc2v165ls7kpp1ghds77qr9jirikalf0-python2.7-ninja-1.11.1.drv' failed to build
error: 1 dependencies of derivation '/nix/store/6pzzyhf0q3hi8im0pbq2l9mzvc05x9n5-python2.7-pycairo-1.18.2.drv' failed to build
error: 1 dependencies of derivation '/nix/store/w7c5pp51sax64b9b8j8xd9272l8z73pl-python-2.7.18.6-env.drv' failed to build
error: 1 dependencies of derivation '/nix/store/cci2flclhx9fw9zb3afdkg5j4r1b78zj-gimp-2.10.34.drv' failed to build
error: 1 dependencies of derivation '/nix/store/n0kvgsdzjm4p1x0ll7is5ax84rnqhicr-user-drv-overlays.drv' failed to build
```
However, the commit history of `flit-core` is rather benign, and I cannot figure out why it failed; after traversing the dependency tree it seems that this commit seems like the only major change that might contribute to the failure. I am no expert in python though; so this might be wrong...

bryango commented:
> > I believe python2 is broken by this PR; it lands on master via #248496, and I can get python2 to build on [its parent](https://github.com/NixOS/nixpkgs/commit/bceafd3a3ad0494c87922993a271606acabccf3d) but afterward, not anymore. Maybe it's time to give up python2 and mark it as broken...
> 
> It builds fine on both master and staging-next.

Ah my bad, it should be `python2.7-flit-core` which is a dependency of `python-2.7.18.6-env`, I missed the `env` part... Sorry for the misleading information...

tjni commented:
This makes sense to me! I'll see what I can do to fix it.

bryango commented:
I think I roughly understand what happened now: the only reason that reverting this PR works for me is because here `flit-core` is introduced as a build dependency. It may be that `python2.7-flit-core` has been failing forever, but I wouldn't notice that because it never entered my dependency tree before. So yeah, the culprit should be `flit-core` all along, but I dismissed it too early... So ninja-python is fine on its own!

tjni commented:
I opened https://github.com/NixOS/nixpkgs/pull/251548 to undo this change on Python 2 and buy myself some time to think about how to make this better.

bryango commented:
Wow, awesome! Thank you very much for the help!
